modules/core/04-channel/keeper/upgrade.go

```diff
@@ -95,6 +96,188 @@ func (k Keeper) WriteUpgradeInitChannel(
+	}
+
+	if proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
+		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannel, "proposed channel upgrade is invalid")
+	}
+
+	if !channel.Ordering.SubsetOf(proposedUpgradeChannel.Ordering) {
+		return 0, "", errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
+	}
+
+	if counterpartyChannel.Ordering != proposedUpgradeChannel.Ordering {
+		return 0, "", errorsmod.Wrapf(types.ErrInvalidChannelOrdering, "channel ordering of counterparty channel and proposed channel must be equal")
+	}
+
+	connectionEnd, err := k.GetConnection(ctx, proposedUpgradeChannel.ConnectionHops[0])
+	if err != nil {
+		return 0, "", err
+	}
+
+	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
+		return 0, "", errorsmod.Wrapf(
+			connectiontypes.ErrInvalidConnectionState,
+			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
+		)
+	}
+
+	if connectionEnd.GetCounterparty().GetConnectionID() != counterpartyChannel.ConnectionHops[0] {
+		return 0, "", errorsmod.Wrapf(connectiontypes.ErrInvalidConnection, "unexpected counterparty channel connection hops, expected %s but got %s", connectionEnd.GetCounterparty().GetConnectionID(), counterpartyChannel.ConnectionHops[0])
+	}
+
+	if err := k.connectionKeeper.VerifyChannelState(ctx, connectionEnd, proofHeight, proofChannel, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, counterpartyChannel); err != nil {
+		return 0, "", err
+	}
+
+	upgradeTimeout := types.UpgradeTimeout{TimeoutHeight: timeoutHeight, TimeoutTimestamp: timeoutTimestamp}
+	if err := k.connectionKeeper.VerifyChannelUpgradeTimeout(ctx, connectionEnd, proofHeight, proofUpgradeTimeout, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, upgradeTimeout); err != nil {
+		return 0, "", err
+	}
+
+	if err := k.connectionKeeper.VerifyChannelUpgradeSequence(ctx, connectionEnd, proofHeight, proofUpgradeSequence, proposedUpgradeChannel.Counterparty.PortId,
+		proposedUpgradeChannel.Counterparty.ChannelId, counterpartyUpgradeSequence); err != nil {
+		return 0, "", err
+	}
+
+	// check if upgrade timed out by comparing it with the latest height of the chain
+	selfHeight := clienttypes.GetSelfHeight(ctx)
+	if !timeoutHeight.IsZero() && selfHeight.GTE(timeoutHeight) {
+		return 0, "", k.RestoreChannel(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout)
+	}
+
+	// check if upgrade timed out by comparing it with the latest timestamp of the chain
+	if timeoutTimestamp != 0 && uint64(ctx.BlockTime().UnixNano()) >= timeoutTimestamp {
+		return 0, "", k.RestoreChannel(ctx, portID, channelID, upgradeSequence, types.ErrUpgradeTimeout)
+	}
```

Comment:
Should we just reject the transaction here instead of writing an error receipt? Opened an issue on the spec https://github.com/cosmos/ibc/issues/953

Reply:
from @AdityaSripal's [reply](https://github.com/cosmos/ibc/issues/953#issuecomment-1494210735) `I write an error receipt here to support the cancellation case, but we could remove it and only rely on timeout.`